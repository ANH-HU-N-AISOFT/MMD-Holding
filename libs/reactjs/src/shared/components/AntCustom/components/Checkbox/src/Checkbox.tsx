import { Checkbox as AntCheckbox, CheckboxProps as AntCheckboxProps } from 'antd';
import classNames from 'classnames';
import { FC, useState } from 'react';
import { useDeepCompareEffect, useIsMounted } from '../../../../../hooks';
import { useInitializeContext } from '../../../base';

export interface Props
  extends Pick<AntCheckboxProps, 'children' | 'className' | 'checked' | 'disabled' | 'indeterminate'> {
  /**
   * Callback function that is triggered when the checkbox state changes.
   *
   * @param {boolean} value - The new checked state of the checkbox.
   */
  onChange?: (value: boolean) => void;
}

/**
 * Checkbox component that extends the functionality of the Ant Design Checkbox component
 * by providing additional customization and support for stricter type safety.
 * It maintains its own state and triggers a callback function when its state changes.
 *
 * @param {Props} props - The properties for the Checkbox component.
 * @param {ReactNode} [props.children] - Content to be displayed next to the checkbox.
 * @param {string} [props.className] - Custom CSS class for styling the checkbox.
 * @param {boolean} [props.checked] - Whether the checkbox is checked.
 * @param {boolean} [props.disabled] - Whether the checkbox is disabled.
 * @param {boolean} [props.indeterminate] - Whether the checkbox is indeterminate.
 * @param {Function} [props.onChange] - Callback function triggered when the checkbox state changes.
 * @returns {ReactNode} The rendered Checkbox component.
 */
export const Checkbox: FC<Props> = ({ children, className, disabled, indeterminate, checked, onChange }) => {
  useInitializeContext();
  const isMounted = useIsMounted();
  const [checkedState, setCheckedState] = useState(checked);

  const handleChange: AntCheckboxProps['onChange'] = event => {
    setCheckedState(event.target.checked);
    onChange?.(event.target.checked);
  };

  useDeepCompareEffect(() => {
    setCheckedState(checked);
  }, [checked]);

  return (
    <AntCheckbox
      children={children}
      className={classNames('Checkbox__container', className)}
      disabled={disabled}
      indeterminate={indeterminate}
      checked={isMounted ? checkedState : undefined}
      onChange={handleChange}
    />
  );
};
