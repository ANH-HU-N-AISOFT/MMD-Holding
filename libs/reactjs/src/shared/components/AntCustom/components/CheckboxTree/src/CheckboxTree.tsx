import { Tree as AntTree, TreeProps as AntTreeProps } from 'antd';
import { SwitcherIcon } from 'antd/es/tree/Tree';
import classNames from 'classnames';
import { isEmpty } from 'ramda';
import { ReactNode, useMemo } from 'react';
import { useInitializeContext } from '../../../base';
import { Leaf } from './types/Leaf';
import { TreeDataNode } from './types/TreeDataNode';
import { leavesToAntTreeData } from './utils/leavesToAntTreeData';

export interface Props<Model>
  extends Pick<AntTreeProps, 'className' | 'checkedKeys' | 'disabled' | 'height' | 'defaultExpandAll'> {
  /** The array of leaf nodes that make up the tree data */
  data: Leaf<Model>[];
  /** Callback function triggered when a node's checked state changes. */
  onCheck?: (
    keys: undefined | string[],
    options: undefined | Array<Omit<Leaf<Model>, 'disabled' | 'hidden' | 'parent'>>,
  ) => void;
  /** Customize expand/collapse icons for tree nodes (With default rotate angular style). */
  iconExpand?: (props: { expanded: boolean }) => ReactNode;
  /** The keys of the currently expanded nodes. */
  expandedKeys?: string[];
  /** Callback function triggered when a node's expanded state changes. */
  onExpand?: (keys: string[]) => void;
}

/**
 * CheckboxTree component that extends the functionality of the Ant Design Tree component
 * by providing additional customization and support for stricter type safety.
 *
 * @template Model - The type of the raw data associated with each leaf.
 * @param {Props<Model>} props - The props for the CheckboxTree component.
 * @param {string} [props.className] - The CSS class for styling the tree component.
 * @param {Leaf<Model>[]} props.data - The array of leaf nodes that make up the tree data.
 * @param {string[]} [props.checkedKeys] - The keys of the currently checked nodes.
 * @param {boolean} [props.disabled] - Whether the entire tree is disabled.
 * @param {number} [props.height] - The height of the tree component.
 * @param {number} [props.defaultExpandAll] - Whether to expand all treeNodes by default.
 * @param {Function} [props.onCheck] - Callback function triggered when a node's checked state changes.
 * @param {Function} [props.iconExpand] - Customize expand/collapse icons for tree nodes (With default rotate angular style).
 * @param {Function} [props.expandedKeys] - The keys of the currently expanded nodes.
 * @param {Function} [props.onExpand] - Callback function triggered when a node's expanded state changes.
 * @returns {ReactNode} The rendered CheckboxTree component.
 */
export const CheckboxTree = <Model,>({
  className,
  data,
  checkedKeys,
  disabled,
  height,
  onCheck,
  iconExpand,
  expandedKeys,
  onExpand,
  defaultExpandAll = true,
}: Props<Model>): ReactNode => {
  useInitializeContext();

  const treeData = useMemo(() => {
    return leavesToAntTreeData(data);
  }, [data]);

  const handleCheck: AntTreeProps<TreeDataNode>['onCheck'] = (keys, info) => {
    const isUndefined = isEmpty(keys) || null;
    onCheck?.(
      isUndefined ? undefined : (keys as string[]),
      isUndefined
        ? undefined
        : info.checkedNodes.map(node => ({
            label: node.checkboxLabel,
            rawData: node.rawData,
            value: node.checkboxValue,
            isLeaf: node.isLeaf,
          })),
    );
  };

  const switcherIcon: SwitcherIcon = ({ expanded }) => {
    return iconExpand?.({ expanded: !!expanded });
  };

  return (
    <AntTree<TreeDataNode>
      // Fixed props
      autoExpandParent
      blockNode
      checkable
      draggable={false}
      focusable
      multiple={false}
      selectable={false}
      showIcon={false}
      showLine={false}
      virtual
      // Customizable props
      defaultExpandAll={defaultExpandAll}
      {...(expandedKeys ? { expandedKeys } : {})}
      onExpand={data => onExpand?.(data as string[])}
      switcherIcon={switcherIcon}
      checkedKeys={checkedKeys}
      disabled={disabled}
      className={classNames('CheckboxTree__container', className)}
      height={height}
      treeData={treeData}
      onCheck={handleCheck}
      fieldNames={{
        title: 'checkboxLabel',
        key: 'checkboxValue',
      }}
    />
  );
};
