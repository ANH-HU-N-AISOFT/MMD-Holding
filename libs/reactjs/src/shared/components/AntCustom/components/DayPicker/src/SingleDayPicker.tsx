import classNames from 'classnames';
import { isEmpty } from 'ramda';
import { FC, ReactNode, useMemo, useState } from 'react';
import { Dayjs } from 'utilities';
import { useDeepCompareEffect, useIsMounted } from '../../../../../hooks';
import './css/SingleDayPicker.css';
import { useInitializeContext } from '../../../base';
import { AntDatePicker, AntDatePickerProps } from '../../../base/AntDatePicker';
import { Format } from './types/Format';
import { detectTimeComponents } from './utils/detectTimeComponents';

export interface Props
  extends Pick<
    AntDatePickerProps,
    'className' | 'allowClear' | 'disabled' | 'showNow' | 'suffixIcon' | 'locale' | 'onBlur' | 'onFocus'
  > {
  /** Function to specify the dates that should be disabled */
  disabledDate?: (date: Dayjs) => boolean;
  /** A function that returns an array of disabled hours. */
  disabledHours?: () => number[];
  /**  A function that returns an array of disabled minutes based on the provided hour. */
  disabledMinutes?: (params: { hours: number }) => number[];
  /** A function that returns an array of disabled seconds based on the provided hour and minute. */
  disabledSeconds?: (params: { hours: number; minutes: number }) => number[];
  /** Preset ranges for quick selection. */
  presets?: Array<{ label: ReactNode; value: Dayjs }>;
  /** Placeholder text for the input. */
  placeholder?: string;
  /** The format in which the date and/or time is displayed. */
  format?: Format;
  /** The currently selected date and time value. */
  value?: Dayjs;
  /** A function to handle changes to the selected date and time value. */
  onChange?: (value: Dayjs | undefined) => void;
}

/**
 * `SingleDayPicker` is a component that allows users to select a single day, with optional time components.
 * It provides various customization options such as disabling specific hours, minutes, and seconds.
 *
 * @param {Props} props - The properties for the RangeDayPicker component.
 * @param {string} [props.className] - Custom CSS class for the date picker.
 * @param {boolean} [props.allowClear] - Whether to show a clear button.
 * @param {string} [props.placeholder] - Placeholder text for the input.
 * @param {boolean} [props.disabled] - Whether the date picker is disabled.
 * @param {Object[]} [props.presets] - Preset ranges for quick selection.
 * @param {boolean} [props.showNow] - Whether to show the "Now" button.
 * @param {Function} [props.disabledDate] - Function to specify the dates that should be disabled.
 * @param {ReactNode} [props.suffixIcon] - Custom suffix icon for the date picker.
 * @param {Object} [props.locale] - Locale configuration for the date picker.
 * @param {Function} [props.disabledHours] - Function to specify the hours that should be disabled.
 * @param {Function} [props.disabledMinutes] - Function to specify the minutes that should be disabled.
 * @param {Function} [props.disabledSeconds] - Function to specify the seconds that should be disabled.
 * @param {Format} [props.format] - Format for displaying the date and time.
 * @param {[Dayjs, Dayjs]} [props.value] - Current value of the date range picker.
 * @param {Function} [props.onChange] - Callback function triggered when the selected date range changes.
 * @param {Function} [props.onBlur] - Callback function triggered when blur the date picker.
 * @param {Function} [props.onFocus] - Callback function triggered when focus the date picker.
 * @returns {JSX.Element} The rendered `SingleDayPicker` component.
 */
export const SingleDayPicker: FC<Props> = ({
  allowClear = true,
  className,
  disabled,
  placeholder,
  format = 'DD/MM/YYYY',
  presets,
  showNow = true,
  suffixIcon,
  disabledDate,
  disabledHours = (): number[] => [],
  disabledMinutes = (): number[] => [],
  disabledSeconds = (): number[] => [],
  locale,
  onBlur = (): void => undefined,
  onFocus = (): void => undefined,
  onChange = (): void => undefined,
  value,
}: Props) => {
  useInitializeContext();
  const [valueState, setValueState] = useState(value);
  const isMounted = useIsMounted();

  const { hasHour, hasMinute, hasSecond } = useMemo(() => {
    return detectTimeComponents(format);
  }, [format]);

  const handleChange: Props['onChange'] = value => {
    let value_ = value;
    if (hasSecond) {
      value_ = value?.startOf('second');
    } else if (hasMinute) {
      value_ = value?.startOf('minute');
    } else if (hasHour) {
      value_ = value?.startOf('hour');
    } else {
      value_ = value?.startOf('day');
    }

    const isUndefined = isEmpty(value_) || null;
    setValueState(isUndefined ? undefined : value_);
    onChange?.(isUndefined ? undefined : value_);
  };

  useDeepCompareEffect(() => {
    setValueState(value);
  }, [value?.valueOf()]);

  return (
    <AntDatePicker
      onBlur={onBlur}
      onFocus={onFocus}
      needConfirm
      picker="date"
      locale={locale}
      placeholder={placeholder}
      disabled={disabled}
      allowClear={allowClear}
      format={format}
      presets={presets}
      showNow={showNow}
      showTime={{
        showHour: hasHour,
        showMinute: hasMinute,
        showSecond: hasSecond,
        showMillisecond: false,
        disabledTime: () => ({
          disabledHours,
          disabledMinutes: hours => disabledMinutes({ hours }),
          disabledSeconds: (hours, minutes) => disabledSeconds({ hours, minutes }),
        }),
      }}
      disabledDate={disabledDate}
      suffixIcon={suffixIcon}
      popupClassName="SingleDayPicker__popup"
      className={classNames('SingleDayPicker__container', className)}
      onChange={handleChange}
      value={isMounted ? valueState : undefined}
    />
  );
};
