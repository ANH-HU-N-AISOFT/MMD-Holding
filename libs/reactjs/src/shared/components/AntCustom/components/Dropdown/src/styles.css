.Dropdown__divider {
  margin: 0;
}

.Dropdown__footer {
  padding: var(--yy-padding-xxs);
}
