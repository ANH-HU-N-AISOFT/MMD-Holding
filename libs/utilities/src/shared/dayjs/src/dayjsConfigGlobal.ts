import dayjs, { Dayjs } from 'dayjs';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import relativeTime from 'dayjs/plugin/relativeTime';
import timezone from 'dayjs/plugin/timezone';
import updateLocale from 'dayjs/plugin/updateLocale';
import utc from 'dayjs/plugin/utc';
import { isBrowser } from '../../isBrowser';

dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(relativeTime);
dayjs.extend(updateLocale);
dayjs.extend(customParseFormat);

/**
 * Extends the functionality of the Day.js library and provides an overridden method for client-side usage.
 *
 * @param {dayjs.ConfigType} input - The date, string, or timestamp to be parsed or manipulated.
 * @param {dayjs.OptionType} format - The format of the input, if applicable.
 * @param {string} locale - The locale to be used for formatting.
 * @param {boolean} strict - Whether to perform strict parsing.
 * @returns {dayjs.Dayjs} - A Day.js object representing the parsed or manipulated date.
 * @throws {Error} Throws an error if used on the server side, providing a warning about potential issues related to timezone offsets.
 *
 * @example
 * const result = dayjsOverride('2024-03-07', 'YYYY-MM-DD');
 * console.log(result.format('MMMM DD, YYYY')); // Outputs: "March 07, 2024"
 */
const dayjsOverride = (
  input?: dayjs.ConfigType,
  format?: dayjs.OptionType,
  locale?: string,
  strict?: boolean,
): dayjs.Dayjs => {
  if (!isBrowser()) {
    throw new Error(`⚠️ Warning: Using Day.js on the server side may lead to issues related to timezone offsets. Servers typically deal with timestamps and forward them to other services. Mutating or altering date objects with Day.js in a server environment can result in unexpected behavior.

    For server-side operations, consider using libraries that are specifically designed for server environments and timestamp handling. Day.js is best suited for client-side tasks where timezone-related issues are typically less problematic.

    Always ensure that you handle date and time operations in a way that aligns with your server's requirements and avoids unnecessary mutations or alterations of date objects.

    If you are working in a server environment, it's recommended to use server-friendly libraries for date and time handling to prevent unintended complications.
    `);
  }

  return dayjs.apply(this, [input, format, locale, strict]);
};

export type { ConfigType } from 'dayjs';
export { Dayjs, dayjsOverride as dayjs };
