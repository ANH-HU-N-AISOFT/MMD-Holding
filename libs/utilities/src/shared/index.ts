export * from './dayjs';
export * from './delay';
export * from './fetchApi';
export * from './formDataUtils';
export * from './isBrowser';
export * from './localStorage';
export * from './pluralize';
export * from './regexes';
export * from './retry';
export * from './updateURLSearchParamsOfBrowserWithoutNavigation';
export * from './urlSearchParamsUtils';
