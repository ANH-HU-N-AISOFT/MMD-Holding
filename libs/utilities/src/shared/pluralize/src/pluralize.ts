interface Pluralize {
  singular: string;
  plural: string;
  count: number;
}

export const pluralize = ({ count, plural, singular }: Pluralize): string => {
  if (count >= 2) {
    return plural;
  }
  return singular;
};
