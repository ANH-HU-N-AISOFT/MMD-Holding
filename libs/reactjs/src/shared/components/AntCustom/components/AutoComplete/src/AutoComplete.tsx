import { AutoComplete as AntAutoComplete, AutoCompleteProps as AntAutoCompleteProps, Spin } from 'antd';
import classNames from 'classnames';
import { isEmpty } from 'ramda';
import { FC, ReactNode, useMemo, useState } from 'react';
import { useDeepCompareEffect } from '../../../../../hooks';
import './styles.css';
import { useInitializeContext } from '../../../base';
import { Option } from './types/Option';
import { OptionValueType } from './types/OptionValueType';
import { baseFilterOption } from './utils/baseFilterOption';

export interface Props<ValueType extends OptionValueType, RawData = any>
  extends Pick<
    AntAutoCompleteProps<ValueType, Option<ValueType, RawData>>,
    | 'allowClear'
    | 'disabled'
    | 'placeholder'
    | 'className'
    | 'notFoundContent'
    | 'filterOption'
    | 'onBlur'
    | 'onFocus'
    | 'searchValue'
    | 'open'
    | 'onSearch'
    | 'onDropdownVisibleChange'
  > {
  /** The current input value. */
  value?: ValueType;
  /** Callback function that is triggered when the input value changes. */
  onChange?: (value: ValueType | undefined, option: Option<ValueType, RawData> | undefined) => void;
  /** Whether the component is in a loading state, showing a loading indicator. */
  loading?: boolean;
  /** Array of options to be displayed in the dropdown menu. */
  options?: Option<ValueType, RawData>[];
}

/**
 * AutoComplete component that provides an input field with dropdown suggestions based on user input.
 *
 * @template ValueType - The type of the value for each option.
 * @template RawData - The raw data type for each option.
 *
 * @param {Props<ValueType, RawData>} props - The properties for the AutoComplete component.
 * @param {boolean} [props.allowClear=true] - Whether to show a clear button allowing the user to clear the input.
 * @param {boolean} [props.disabled] - Whether the AutoComplete component is disabled.
 * @param {string} [props.placeholder] - Placeholder text to display when the input is empty.
 * @param {string} [props.className] - Custom CSS class for styling the component.
 * @param {ReactNode} [props.notFoundContent] - Content to display when no options match the input.
 * @param {boolean} [props.loading=false] - Whether the component is in a loading state.
 * @param {string} [props.value] - The current value of the input.
 * @param {(value: string | undefined, option: Option<ValueType, RawData> | undefined) => void} [props.onChange] - Callback function that is triggered when the input value changes.
 * @param {boolean | ((inputValue: string, option: Option<ValueType, RawData>) => boolean)} [props.filterOption=baseFilterOption] - Custom filter function to determine whether an option should be shown in the dropdown.
 * @param {Option<ValueType, RawData>[]} [props.options=[]] - Array of options to be displayed in the dropdown menu.
 * @param {boolean} [props.open] - Whether the dropdown menu is open.
 * @param {string} [props.searchValue] - The value of the search input.
 * @param {Function} [props.onDropdownVisibleChange] - Callback function that is triggered when the dropdown visibility changes.
 * @param {Function} [props.onSearch] - Callback function that is triggered when the search input value changes.
 * @returns {ReactNode} The rendered AutoComplete component.
 */
export const AutoComplete = <ValueType extends OptionValueType, RawData = any>({
  allowClear = true,
  disabled,
  placeholder,
  className,
  notFoundContent,
  loading = false,
  value,
  filterOption = baseFilterOption,
  options = [],
  open,
  searchValue,
  onChange,
  onBlur,
  onDropdownVisibleChange,
  onFocus,
  onSearch,
}: Props<ValueType, RawData>): ReactNode => {
  useInitializeContext();
  const [valueState, setValueState] = useState(value);

  const handleChange: Props<ValueType>['onChange'] = (value, option) => {
    const isUndefined = isEmpty(value) || null;
    setValueState(isUndefined ? undefined : value);
    onChange?.(isUndefined ? undefined : value, isUndefined ? undefined : option);
  };

  useDeepCompareEffect(() => {
    setValueState(value);
  }, [value]);

  const options_ = useMemo(() => {
    return options.filter(item => !item.hidden);
  }, [options]);

  const renderLoadingAtDropdown: FC = () => {
    return (
      <div className="AutoComplete__loading">
        <Spin size="large" />
      </div>
    );
  };

  return (
    <AntAutoComplete
      filterOption={filterOption}
      options={options_}
      notFoundContent={loading ? renderLoadingAtDropdown({}) : notFoundContent}
      value={valueState}
      onChange={handleChange as AntAutoCompleteProps<ValueType, Option<ValueType, RawData>>['onChange']}
      allowClear={allowClear}
      disabled={disabled}
      placeholder={placeholder}
      className={classNames('AutoComplete__container', className)}
      open={open}
      searchValue={searchValue}
      onBlur={onBlur}
      onDropdownVisibleChange={onDropdownVisibleChange}
      onFocus={onFocus}
      onSearch={onSearch}
    />
  );
};
