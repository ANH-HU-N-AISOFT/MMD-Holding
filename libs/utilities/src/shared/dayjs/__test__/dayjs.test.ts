// Import the necessary modules for testing
import { dayjs } from '../src/dayjsConfigGlobal';

describe('dayjs', () => {
  it('should return a formatted date for client-side usage', () => {
    const result = dayjs('2024-03-07', 'YYYY-MM-DD');
    expect(result.format('MMMM DD, YYYY')).toBe('March 07, 2024');
  });

  it('should throw an error with a warning for server-side usage', () => {
    // @ts-ignore
    delete global.window;
    expect(() => dayjs('2024-03-07', 'YYYY-MM-DD')).toThrowErrorMatchingSnapshot();
  });
});
