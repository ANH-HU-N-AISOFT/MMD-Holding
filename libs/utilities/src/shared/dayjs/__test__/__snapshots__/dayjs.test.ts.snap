// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dayjs should throw an error with a warning for server-side usage 1`] = `
"⚠️ Warning: Using Day.js on the server side may lead to issues related to timezone offsets. Servers typically deal with timestamps and forward them to other services. Mutating or altering date objects with Day.js in a server environment can result in unexpected behavior.

    For server-side operations, consider using libraries that are specifically designed for server environments and timestamp handling. Day.js is best suited for client-side tasks where timezone-related issues are typically less problematic.

    Always ensure that you handle date and time operations in a way that aligns with your server's requirements and avoids unnecessary mutations or alterations of date objects.

    If you are working in a server environment, it's recommended to use server-friendly libraries for date and time handling to prevent unintended complications.
    "
`;
